pkg-r/R/chat_append.R

```diff
@@ -0,0 +1,589 @@
+    # rendering method in markdown-stream.ts
+    msg_content <- sprintf(
+      "\n\n````````{=html}\n%s\n````````\n\n",
+      msg_content
+    )
+  }
+
+  list(
+    content = msg_content,
+    contentType = content_type,
+    html_deps = ui[["deps"]]
+  )
+}
+
+chat_append_stream <- function(
+  id,
+  stream,
+  role = "assistant",
+  icon = NULL,
+  session = getDefaultReactiveDomain()
+) {
+  result <- chat_append_stream_impl(id, stream, role, icon, session)
+  result <- chat_update_bookmark(id, result, session = session)
+  # Handle erroneous result...
+  result <- promises::catch(result, function(reason) {
+    # ...but rethrow the error as a silent error, so the caller can also handle
+    # it if they want, but it won't bring down the app.
+    class(reason) <- c("shiny.silent.error", class(reason))
+    cnd_signal(reason)
+  })
+
+  promises::catch(result, function(reason) {
+    chat_append_message(
+      id,
+      list(
+        role = role,
+        content = sanitized_chat_error(reason)
+      ),
+      chunk = "end",
+      operation = "append",
+      session = session
+    )
+    rlang::warn(
+      sprintf(
+        "ERROR: An error occurred in `chat_append_stream(id=\"%s\")`",
+        session$ns(id)
+      ),
+      parent = reason
+    )
+  })
+
+  # Note that we're not returning the result of `promises::catch()`, because we
+  # want to return a rejected promise so the caller can see the error. But we
+  # use the `catch()` both to make the error visible to the user *and* to ensure
+  # there's no "unhandled promise error" warning if the caller chooses not to do
+  # anything with it.
+  result
+}
+
+chat_event_enable_input <- function(id, session = getDefaultReactiveDomain()) {
+  session$sendCustomMessage(
+    "shinyChatMessage",
+    list(
+      id = resolve_id(id, session),
+      handler = "shiny-chat-input-enable",
+      obj = NULL
+    )
+  )
+
+  invisible(NULL)
+}
+
+utils:::globalVariables(c("generator_env", "exits", "yield"))
+
+chat_append_stream_impl <- NULL
+rlang::on_load(
+  chat_append_stream_impl <- coro::async(function(
+    id,
+    stream,
+    role = "assistant",
+    icon = NULL,
+    session = shiny::getDefaultReactiveDomain()
+  ) {
+    chat_append_ <- function(content, chunk = TRUE, ...) {
+      chat_append_message(
+        id,
+        msg = list(role = role, content = content),
+        operation = "append",
+        chunk = chunk,
+        session = session,
+        ...
+      )
+    }
+    on.exit(chat_event_enable_input(id, session), add = TRUE)
+
+    res <- fastmap::fastqueue(200)
+
+    do_start_stream <- TRUE
+    this_stream_stage <- stream_stage_state_machine()
+
+    for (msg in stream) {
+      if (promises::is.promising(msg)) {
+        msg <- await(msg)
+      }
+      if (coro::is_exhausted(msg)) {
+        break
+      }
+
+      if (do_start_stream) {
+        chat_append_("", chunk = "start", icon = icon)
+        do_start_stream <- FALSE
+      }
```

Comment:
We delay the first start message until we start getting streaming results back. This keeps the loading message in the chat until the actual streaming begins. (Previously we'd send the first message immediately.)